### DISCUSSION TITLE
ipAddress shenanigans - edge runtime - ratelimit - ipAddress changing between request

### DISCUSSION THREAD

**Discussion by eztinkerdreams** [vercel/community#4657] [Help]
### Summary

IP address from browser request is different on OPTION/GET methods, client ip address changing on multiple request to same endpoint for web request. Doesn't do this for API request from non browser based clients.

### Steps to Reproduce
I am not using a proxy for the request from my local machine to the deployed vercel app.

Ok... so I'm trying to follow [@vercel/edge guid to implement rate limit middleware](https://vercel.com/guides/rate-limiting-edge-middleware-vercel-kv#step-3-adding-@upstash/rate-limit)

I am implementing this code in [solid-start middleware](https://start.solidjs.com/advanced/middleware)
I am using `SSR` Server Side Rending. 

I have `/api` routes, as well as normal page routes `/` for the default page.
The rate limit works just great on the `/api` routes because when I use a tool like the [insomnia rest client](https://insomnia.rest/) The ip address that I get in my vercel logs is typically the same between request

I get the expected 429 response etc...

However... when I make a request to my default url, the rate limiting doesn't work.
In one case, it works, but always just blocks the OPTIONS request, never the GET request.

I put in code to ignore the OPTIONS request, but it still does not work as expected.

The reason is the at the ipAddress that is from the OPTIONS request, and the GET request are different...
In addition making multiple request, i.e. spamming the reload button on the browser causes me to get new IP address each time.
You can see my my `@vercel/kv` the records that were created by me just clicking reload a few times. 
So... what is going on here?

![2023-10-29_05-49-05](https://github.com/vercel/community/assets/7387907/25a9a4e6-8d5f-4cde-86b2-970157ca0c00)

![2023-10-29_05-46-40](https://github.com/vercel/community/assets/7387907/04dd6eb1-dd08-4bbe-988b-c5483a554557)

```
import { kv } from '@vercel/kv';
import { ipAddress } from '@vercel/edge'
import { Ratelimit } from '@upstash/ratelimit';
import { StatusCodes } from 'http-status-codes';

/**
 * Our default rateLimit configuration
 * Ref: https://github.com/upstash/ratelimit
 */
const upstash = new Ratelimit({
  redis: kv,
  analytics: true,
  // 5 requests from the same IP in 10 seconds
  limiter: Ratelimit.slidingWindow(5, '10 s'),
});

// Define which routes you want to rate limit
export const config = {
  matcher: '/',
};

/**
 * This middleware is responsible for adding rate limiting to all routes
 * it will automatically block bad actors who are abusing the app
 * https://vercel.com/guides/rate-limiting-edge-middleware-vercel-kv#step-3-adding-@upstash/rate-limit
 * 
 */
const rateLimit = ({ forward }) => {
  return async (event) => {
    const { clientAddress, request } = event;
    const forwardedFor = request.headers.get("x-forwarded-for")?.split(',').shift();

    console.info(`x-forwarded-for ${forwardedFor} & clientAddress: ${clientAddress}`)
    const ip = ipAddress(request) || forwardedFor || clientAddress || '127.0.0.1';
    const { success, pending, limit, reset, remaining } = await upstash.limit(`client_ip_${ip}`);
    await pending;
    return success ? forward(event) : new Response(null, {
      status: StatusCodes.TOO_MANY_REQUESTS,
      headers: {
        'X-RateLimit-Limit': limit.toString(),
        'X-RateLimit-Reset': reset.toString(),
        'X-RateLimit-Remaining': remaining.toString(),
      }
    });
  };
}

export {
  rateLimit,
}
```

**Comment by eztinkerdreams** [comment]
Update, this works as expected when deployed to a `production` build, rather than a `preview`... Maybe someone can explain why?

**Reply by itisnas** [reply]
Im having the same issue with nextjs app directory (Server side rendering). The ip address seems to be changing initially. Did you manage to solve this?

**Reply by eztinkerdreams** [reply]
No, but I only had the problem in a preview environment where the IP changed on every request from the same client. Upon deploying the same code to production, I did not have the same issue. The IP would not change on every request from the same client. No explanation why. I suspect maybe there is some reverse proxy in play. I have my preview environment setup to require that the user is one of my vercel account users in the preview environment.
